File: a_cmd/ssh.py
"""aio ssh - SSH management (RFC 5322 .txt storage)"""
import sys, os, subprocess as sp, re, shutil, base64
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor as TP
from . _common import _up, _die, DATA_DIR
from .sync import sync, SYNC_ROOT

SSH_DIR = SYNC_ROOT / 'ssh'
def _parse(f): return {k.strip(): v.strip() for line in f.read_text().splitlines() if ':' in line for k, v in [line.split(':', 1)]}
def _save(n, h, pw=None, **kw):
    SSH_DIR.mkdir(parents=True, exist_ok=True); d = _parse(SSH_DIR/f'{n}.txt') if (SSH_DIR/f'{n}.txt').exists() else {}
    d.update({'Name': n, 'Host': h}); pw and d.update({'Password': pw}); d.update({k: v for k, v in kw.items() if v})
    (SSH_DIR/f'{n}.txt').write_text('\n'.join(f"{k}: {v}" for k, v in d.items() if v) + '\n'); sync('ssh')
def _load():
    SSH_DIR.mkdir(parents=True, exist_ok=True); sync('ssh')
    return [d for f in sorted(SSH_DIR.glob('*.txt')) if (d := _parse(f)).get('Name')]
def _rm(n): (SSH_DIR/f'{n}.txt').unlink(missing_ok=True); sync('ssh')
def _os(): return sp.run('uname -sr 2>/dev/null || echo unknown', shell=True, capture_output=True, text=True).stdout.strip()

def run():
    wda = sys.argv[2] if len(sys.argv) > 2 else None
    def _sshd_running(): return _up('localhost:8022') or _up('localhost:22')
    def _sshd_ip():
        try: import socket as S; s=S.socket(S.AF_INET,S.SOCK_DGRAM); s.connect(("8.8.8.8",80)); ip=s.getsockname()[0]; s.close(); return ip
        except: return '?'
    def _sshd_port(): return 8022 if os.environ.get('TERMUX_VERSION') else 22

    raw = _load(); hosts = [(d.get('Name','?'), d.get('Host')) for d in raw]; hmap = {d['Name']: d['Host'] for d in raw if d.get('Host')}; pwmap = {d['Name']: d.get('Password') for d in raw}; osmap = {d['Name']: d.get('OS') for d in raw}

    if wda == 'start': r = sp.run(['sshd'], capture_output=True, text=True) if os.environ.get('TERMUX_VERSION') else sp.run(['sudo', '/usr/sbin/sshd'], capture_output=True, text=True); print(f"✓ sshd started (port {_sshd_port()})") if r.returncode == 0 or _sshd_running() else print(f"x sshd failed: {r.stderr.strip() or 'install openssh-server'}"); return
    if wda == 'stop': sp.run(['pkill', '-x', 'sshd']) if os.environ.get('TERMUX_VERSION') else sp.run(['sudo', 'pkill', '-x', 'sshd']); print("✓ sshd stopped" if not _sshd_running() else "x failed"); return
    if wda in ('status', 's'): u, ip, p = os.environ.get('USER', 'u0_a'), _sshd_ip(), _sshd_port(); print(f"{'✓ RUNNING' if _sshd_running() else 'x STOPPED'}  ssh {u}@{ip} -p {p}"); return
    if not wda: u, ip, p = os.environ.get('USER', 'u0_a'), _sshd_ip(), _sshd_port(); me = next((n for n,h in hosts if ip in h), None); url = sp.run(['git','-C',str(SSH_DIR),'remote','get-url','origin'], capture_output=True, text=True).stdout.strip(); shutil.which('ssh') or print("! pkg install openssh"); print(f"SSH {'✓ ON' if _sshd_running() else 'x OFF'}  →  ssh {u}@{ip} -p {p}{' ('+me+')' if me else ' [not in list]'}\n  {SSH_DIR}\n  {url}\n  start/stop/status  server control\n  setup              install & configure\n  self [name]        add this device\n  add/rm/mv/pw <#>   manage hosts\n  all \"cmd\"          run cmd on all hosts\n  Termux: run 'passwd' to set password before 'a ssh self'\nHosts:"); up=list(TP(8).map(_up,[h or '' for _,h in hosts])); [print(f"  {i}. x {n}: missing Name:/Host:") if not h else print(f"  {i}. {'✓x'[not up[i]]} {n}: {h}{' [pw]' if pwmap.get(n) else ''}{' <- this' if n==me else ''}") for i,(n,h) in enumerate(hosts)] or print("  (none)"); (any(not x for x in up) and not sp.run("ping -c1 -W1 $(ip route|awk '/default/{print $3}')",shell=True,capture_output=True).returncode) and print("! Unreachable hosts - check AP isolation (wired↔wifi blocked)"); print(f"\nAI: a ssh <#|name> cmd"); return
    if wda == 'setup': ip = _sshd_ip(); u = os.environ.get('USER', 'user'); ok = _sshd_running(); cmd = 'pkg install -y openssh && sshd' if os.environ.get('TERMUX_VERSION') else 'sudo apt install -y openssh-server && sudo systemctl enable --now ssh'; (not ok and input("SSH not running. Install? (y/n): ").lower() in ['y', 'yes'] and sp.run(cmd, shell=True)); ok = ok or _sshd_running(); print(f"This: {os.environ.get('HOSTNAME', 'host')} ({u}@{ip}:{_sshd_port()})\nSSH: {'✓ running' if ok else 'x not running'}\n\nTo connect here from another device:\n  ssh {u}@{ip} -p {_sshd_port()}"); return
    if wda == 'key': kf = Path.home()/'.ssh/id_ed25519'; kf.exists() or sp.run(['ssh-keygen','-t','ed25519','-N','','-f',str(kf)]); print(f"Public key:\n{(kf.with_suffix('.pub')).read_text().strip()}"); return
    if wda == 'auth': d = Path.home()/'.ssh'; d.mkdir(exist_ok=True); af = d/'authorized_keys'; k = input("Paste public key: ").strip(); af.open('a').write(f"\n{k}\n"); af.chmod(0o600); print("✓ Added"); return
    if wda=='push-auth': t=sp.run(['gh','auth','token'],capture_output=True,text=True).stdout.strip(); [sp.run([sys.executable,__file__.replace('a_cmd/ssh.py','a.py'),'ssh',sys.argv[3],f"mkdir -p ~/.config/{d}&&echo {base64.b64encode((Path.home()/'.config'/d/f).read_bytes()).decode()}|base64 -d>~/.config/{d}/{f}"+";touch ~/.local/share/a/.auth_shared"*(d<'q')]) for d,f in[('rclone','rclone.conf'),('gh','hosts.yml')]]; t and sp.run([sys.executable,__file__.replace('a_cmd/ssh.py','a.py'),'ssh',sys.argv[3],f'echo "{t}"|gh auth login --with-token']); print("✓"); return
    if wda == 'self':
        wsl = os.path.exists('/proc/version') and 'microsoft' in open('/proc/version').read().lower()
        # === WSL: needs Windows port forward (2222->22) to be reachable from LAN ===
        if wsl:
            wsl_ip = sp.run("hostname -I", shell=True, capture_output=True, text=True).stdout.split()[0]
            win_ip = re.search(r'192\.168\.1\.\d+', sp.run(["powershell.exe","-c","ipconfig"],capture_output=True,text=True).stdout)
            if not win_ip: _die("x No Windows LAN IP (192.168.1.x) found")
            ip, p = win_ip.group(), 2222; print(f"WSL: {wsl_ip} → Windows: {ip}:{p}")
            sp.run("pgrep -x sshd >/dev/null || sudo service ssh start", shell=True)
            # Check/setup port forward
            pf = sp.run(["powershell.exe","-c","netsh interface portproxy show all"], capture_output=True, text=True).stdout
            if "2222" not in pf:
                print("Setting up Windows port forward (UAC prompt)...")
                ps1 = f'netsh interface portproxy delete v4tov4 listenport=2222 listenaddress=0.0.0.0 2>$null\nnetsh interface portproxy add v4tov4 listenport=2222 listenaddress=0.0.0.0 connectport=22 connectaddress={wsl_ip}\nnetsh advfirewall firewall delete rule name="WSL SSH" 2>$null\nnetsh advfirewall firewall add rule name="WSL SSH" dir=in action=allow protocol=tcp localport=2222\nWrite-Host "Done" -ForegroundColor Green; netsh interface portproxy show all; pause'
                open("/mnt/c/Windows/Temp/wsl_ssh.ps1","w").write(ps1)
                sp.run(["powershell.exe","-c","Start-Process powershell -Verb RunAs -ArgumentList '-ExecutionPolicy','Bypass','-File','C:\\Windows\\Temp\\wsl_ssh.ps1'"])
                input("Press Enter after Windows admin window shows 'Done'...")
                if "2222" not in sp.run(["powershell.exe","-c","netsh interface portproxy show all"], capture_output=True, text=True).stdout: _die("x Port forward failed")
            print("✓ Port forward OK")
        # === macOS: may need to enable Remote Login ===
        elif sys.platform=='darwin':
            sp.run(['nc','-z','localhost','22'],capture_output=True).returncode and (input("Remote Login off. Enable? (y/n): ").lower() in ['y','yes'] and sp.run(['sudo','systemsetup','-setremotelogin','on']) or _die("x enable Remote Login"))
            ip, p = sp.run("ipconfig getifaddr en0", shell=True, capture_output=True, text=True).stdout.strip() or _sshd_ip(), _sshd_port()
        # === Linux/Termux: use local IP ===
        else: ip, p = sp.run("hostname -I 2>/dev/null | awk '{print $1}'", shell=True, capture_output=True, text=True).stdout.strip() or _sshd_ip(), _sshd_port()
        # self [name] [pw]
        u=os.environ.get('USER') or sp.run('whoami',shell=True,capture_output=True,text=True).stdout.strip() or 'user'; h=f"{u}@{ip}"+(f":{p}"if p!=22 else""); a=sys.argv[3:]; n=a[0]if a else u; pw=a[1]if len(a)>1 else input("Pw:").strip()
        os.system(f'sshpass -p "{pw}" ssh -oStrictHostKeyChecking=no -p{p} localhost exit')or 0
        _save(n, h, pw, OS=_os()); print(f"✓ {n}={h} [{_os()}]"); return
    if wda in ('info','i'): [print(f"{n}: ssh {'-p '+hp[1]+' ' if len(hp:=h.rsplit(':',1))>1 else ''}{hp[0]}{' ('+osmap[n]+')' if osmap.get(n) else ''}") for n,h in hosts]; return
    if wda == 'os':
        cmd = 'uname -sr'; results = []
        def _get(d): n,h,pw = d.get('Name'), d.get('Host'), d.get('Password'); hp=h.rsplit(':',1) if h else ['','']; r=sp.run((['sshpass','-p',pw]if pw else[])+['ssh','-oConnectTimeout=5','-oStrictHostKeyChecking=no']+(['-p',hp[1]]if len(hp)>1 else[])+[hp[0],cmd],capture_output=1,text=1); return (n,h,pw,r.stdout.strip() if not r.returncode else None)
        for n,h,pw,os_info in TP(8).map(_get, raw):
            if os_info: _save(n, h, pw, OS=os_info); print(f"✓ {n}: {os_info}")
            else: print(f"x {n}: unreachable")
        return
    if wda in ('all','*') and len(sys.argv)>3:
        cmd='bash -ic '+repr(' '.join(sys.argv[3:])); ok_l, fail_l = [], []
        def _run(nh): n,h=nh; pw=pwmap.get(n); hp=h.rsplit(':',1); r=sp.run((['sshpass','-p',pw]if pw else[])+['ssh','-oConnectTimeout=5','-oStrictHostKeyChecking=no','-oIdentitiesOnly=yes']+(['-p',hp[1]]if len(hp)>1 else[])+[hp[0],cmd],capture_output=1,text=1); return(n,not r.returncode,(r.stdout or r.stderr).strip())
        for n,ok,out in TP(8).map(_run,hosts): (ok_l if ok else fail_l).append(n); print(f"\n{'✓' if ok else 'x'} {n}"); out and print('\n'.join('  '+l for l in out.split('\n')[:100]))
        print(f"\n✓ {len(ok_l)}" + (f"  x {len(fail_l)} ({','.join(fail_l)})" if fail_l else ""))
        return
    if wda == 'rm' and len(sys.argv) > 3: a=sys.argv[3]; n=hosts[int(a)][0] if a.isdigit() and int(a)<len(hosts) else a; _rm(n); print(f"✓ rm {n}"); return
    if wda == 'pw' and len(sys.argv) > 3: a=sys.argv[3]; n=hosts[int(a)][0] if a.isdigit() and int(a)<len(hosts) else a; pw=input(f"Pw for {n}: ").strip(); _save(n, hmap[n], pw); print(f"✓ {n}"); return
    if wda in ('mv','rename') and len(sys.argv) > 4: o,n=sys.argv[3:5]; _rm(o); _save(n, hmap.get(o,""), pwmap.get(o)); print(f"✓ {o} → {n}"); return
    if wda == 'add': h=re.sub(r'\s+-p\s*(\d+)',r':\1',input("Host (user@ip): ").strip()); n=input("Name: ").strip() or h.split('@')[-1].split(':')[0].split('.')[-1]; pw=input("Pw? ").strip() or None; hp=h.rsplit(':',1); 'ok' in sp.run((['sshpass','-p',pw] if pw else [])+['ssh','-o','ConnectTimeout=5','-o','StrictHostKeyChecking=no']+(['-p',hp[1]] if len(hp)>1 else [])+[hp[0],'echo ok'],capture_output=True,text=True).stdout or _die("x auth failed"); _save(n, h, pw); print(f"✓ {n}={h}{' [pw]' if pw else ''}"); return
    nm = hosts[int(wda)][0] if wda.isdigit() and int(wda) < len(hosts) else (_die(f"x No host #{wda}. Run: a ssh") if wda.isdigit() else wda); shutil.which('ssh') or _die("x ssh not installed"); h=hmap.get(nm,nm); pw=pwmap.get(nm); hp=h.rsplit(':',1)
    if len(sys.argv)>3:
        tty = sys.stdout.isatty(); cmd = ['ssh'] + (['-tt'] if tty else ['-oConnectTimeout=10']) + ['-oStrictHostKeyChecking=no'] + (['-p',hp[1]] if len(hp)>1 else []) + [hp[0], 'bash -ic '+repr(' '.join(sys.argv[3:]))+' 2>&1']
        r = sp.run((['sshpass','-p',pw] if pw else [])+cmd, capture_output=not tty, text=True); tty or print('\n'.join(l for l in r.stdout.split('\n') if not l.startswith('bash:'))); return
    cmd=['ssh','-tt','-o','StrictHostKeyChecking=accept-new']+(['-p',hp[1]] if len(hp)>1 else [])+[hp[0]]
    if not pw and nm in hmap: pw=input("Password? ").strip(); pw and _save(nm, hmap[nm], pw)
    pw and not shutil.which('sshpass') and _die("x need sshpass"); print(f"Connecting to {nm}...", file=sys.stderr, flush=True); os.execvp('sshpass',['sshpass','-p',pw]+cmd) if pw else os.execvp('ssh',cmd)

File: lib/ssh.c
/* ── ssh ── */
#define SMUX " -oControlMaster=auto -oControlPath=%%d/.ssh/a-%%C -oControlPersist=300"
static void ssh_parse(const char*h,char*hp,char*port){
    snprintf(hp,256,"%s",h);char*c=strrchr(hp,':');if(c){snprintf(port,8,"%s",c+1);*c=0;}else snprintf(port,8,"22");}
static int ssh_pre(char*c,int sz,const char*pw,const char*opts,const char*port,const char*hp){
    int n=0;if(pw&&pw[0])n=snprintf(c,(size_t)sz,"sshpass -p '%s' ",pw);
    return n+snprintf(c+n,(size_t)(sz-n),"ssh" SMUX " %s -p %s '%s'",opts,port,hp);}
static void ssh_save(const char*dir,const char*n,const char*h,const char*pw){
    char f[P],d[B];snprintf(f,P,"%s/%s.txt",dir,n);
    snprintf(d,B,"Name: %s\nHost: %s\n%s%s%s",n,h,pw&&pw[0]?"Password: ":"",pw?pw:"",pw&&pw[0]?"\n":"");
    writef(f,d);sync_repo();}
static void ssh_savex(const char*dir,const char*n,const char*h,const char*pw,const char*k,const char*v){
    char f[P],d[B];snprintf(f,P,"%s/%s.txt",dir,n);
    int l=snprintf(d,B,"Name: %s\nHost: %s\n",n,h);
    if(pw&&pw[0])l+=snprintf(d+l,(size_t)(B-l),"Password: %s\n",pw);
    if(k&&v&&v[0])snprintf(d+l,(size_t)(B-l),"%s: %s\n",k,v);
    writef(f,d);sync_repo();}
static int cmd_ssh(int argc,char**argv){
    char dir[P];snprintf(dir,P,"%s/ssh",SROOT);mkdirp(dir);sync_bg();
    typedef struct{char name[128],host[256],pw[256];}host_t;
    host_t H[32];int nh=0,arc=0;
    char paths[32][P];int np=listdir(dir,paths,32);
    for(int i=np-1;i>=0&&nh<32;i--){
        kvs_t kv=kvfile(paths[i]);const char*n=kvget(&kv,"Name");if(!n)continue;
        int dup=0;for(int j=0;j<nh;j++)if(!strcmp(H[j].name,n)){dup=1;break;}
        if(dup){do_archive(paths[i]);arc++;continue;}
        snprintf(H[nh].name,128,"%s",n);const char*h=kvget(&kv,"Host"),*p=kvget(&kv,"Password");
        snprintf(H[nh].host,256,"%s",h?h:"");snprintf(H[nh].pw,256,"%s",p?p:"");nh++;}
    if(arc)sync_bg();
    const char*sub=argc>2?argv[2]:NULL;
    /* list */
    if(!sub){int on=!system("pgrep -x sshd >/dev/null 2>&1");
        printf("SSH sshd:%s\n\n",on?" \033[32mon\033[0m":" \033[31moff\033[0m");
        for(int i=0;i<nh;i++){int s=!strcmp(H[i].name,DEV);
            printf("  %d. %s%s%s: %s%s\n",i,s?"\033[32m":"",H[i].name,s?" (self)\033[0m":"",H[i].host,H[i].pw[0]?" [pw]":"");}
        if(!nh)puts("  (none)");
        puts("\na ssh <#|name> [cmd]  add/self/start/stop/all/rm\n"
             "  setup/key/auth/os/info/pw/mv");return 0;}
    /* start/stop/status */
    if(!strcmp(sub,"start")){(void)!system("sshd 2>/dev/null||sudo /usr/sbin/sshd");puts("\xe2\x9c\x93");return 0;}
    if(!strcmp(sub,"stop")){(void)!system("pkill -x sshd 2>/dev/null||sudo pkill -x sshd");puts("\xe2\x9c\x93");return 0;}
    if(!strcmp(sub,"status")||!strcmp(sub,"s")){
        int on=!system("pgrep -x sshd >/dev/null 2>&1");char ip[128];
        pcmd("hostname -I 2>/dev/null|awk '{printf $1}'",ip,128);
        const char*u=getenv("USER");int p=access("/data/data/com.termux",F_OK)?22:8022;
        printf("%s ssh %s@%s -p %d\n",on?"\xe2\x9c\x93":"x",u?u:"",ip,p);return 0;}
    /* setup — install openssh */
    if(!strcmp(sub,"setup")){
        int on=!system("pgrep -x sshd >/dev/null 2>&1");
        if(!on){printf("SSH not running. Install? (y/n): ");char yn[8];
            if(fgets(yn,8,stdin)&&(yn[0]=='y'||yn[0]=='Y')){
                if(!access("/data/data/com.termux",F_OK))(void)!system("pkg install -y openssh && sshd");
                else (void)!system("sudo apt install -y openssh-server && sudo systemctl enable --now ssh");}}
        on=!system("pgrep -x sshd >/dev/null 2>&1");
        printf("SSH: %s\n",on?"\xe2\x9c\x93 running":"x not running");return 0;}
    /* key — generate ed25519 key */
    if(!strcmp(sub,"key")){char kf[P];snprintf(kf,P,"%s/.ssh/id_ed25519",HOME);
        struct stat st;if(stat(kf,&st)){char c[B];snprintf(c,B,"ssh-keygen -t ed25519 -N '' -f '%s'",kf);(void)!system(c);}
        char pub[P];snprintf(pub,P,"%s.pub",kf);catf(pub);return 0;}
    /* auth — add authorized key */
    if(!strcmp(sub,"auth")){char k[B];printf("Paste public key: ");if(!fgets(k,B,stdin))return 1;
        k[strcspn(k,"\n")]=0;char af[P];snprintf(af,P,"%s/.ssh/authorized_keys",HOME);
        char d[P];snprintf(d,P,"%s/.ssh",HOME);mkdirp(d);
        FILE*f=fopen(af,"a");if(f){fprintf(f,"\n%s\n",k);fclose(f);chmod(af,0600);puts("\xe2\x9c\x93");}return 0;}
    /* push-auth — push gh/rclone creds to remote host */
    if(!strcmp(sub,"push-auth")&&argc>3){
        char tok[512];pcmd("gh auth token 2>/dev/null",tok,512);tok[strcspn(tok,"\n")]=0;
        const char*tgt=argv[3];
        /* push rclone.conf */
        {char rc[P],c[B*2];snprintf(rc,P,"%s/.config/rclone/rclone.conf",HOME);
        if(fexists(rc)){snprintf(c,B*2,"base64 '%s'|a ssh %s 'mkdir -p ~/.config/rclone&&base64 -d>~/.config/rclone/rclone.conf'",rc,tgt);(void)!system(c);}}
        /* push gh hosts.yml */
        {char gh[P],c[B*2];snprintf(gh,P,"%s/.config/gh/hosts.yml",HOME);
        if(fexists(gh)){snprintf(c,B*2,"base64 '%s'|a ssh %s 'mkdir -p ~/.config/gh&&base64 -d>~/.config/gh/hosts.yml'",gh,tgt);(void)!system(c);}}
        /* push gh token */
        if(tok[0]){char c[B*2];snprintf(c,B*2,"a ssh %s 'echo \"%s\"|gh auth login --with-token'",tgt,tok);(void)!system(c);}
        puts("\xe2\x9c\x93");return 0;}
    /* add — interactive */
    if(!strcmp(sub,"add")){char h[256],n[128],pw[256];
        printf("Host: ");if(!fgets(h,256,stdin))return 1;h[strcspn(h,"\n")]=0;
        printf("Name: ");if(!fgets(n,128,stdin))return 1;n[strcspn(n,"\n")]=0;
        if(!n[0]){char*at=strchr(h,'@');snprintf(n,128,"%s",at?at+1:h);}
        printf("Password: ");if(!fgets(pw,256,stdin))return 1;pw[strcspn(pw,"\n")]=0;
        /* validate connection */
        {char hp[256],port[8];ssh_parse(h,hp,port);
        char tc[B];int l=ssh_pre(tc,B,pw,"-oConnectTimeout=5 -oStrictHostKeyChecking=no",port,hp);
        snprintf(tc+l,(size_t)(B-l)," 'echo ok' 2>&1");
        char o[64];if(pcmd(tc,o,64)||!strstr(o,"ok")){printf("x auth failed: %s",o);return 1;}}
        ssh_save(dir,n,h,pw);printf("\xe2\x9c\x93 %s\n",n);return 0;}
    /* self — register this device */
    if(!strcmp(sub,"self")){char ip[128]="",port[8]="22",h[256];
        const char*u=getenv("USER");const char*nm=argc>3?argv[3]:DEV;
        /* WSL detection */
        int wsl=0;{char pv[64];pcmd("grep -ci microsoft /proc/version 2>/dev/null",pv,64);wsl=atoi(pv)>0;}
        if(wsl){
            pcmd("powershell.exe -c \"ipconfig\"|grep -oP '192\\.168\\.\\d+\\.\\d+'|head -1",ip,128);ip[strcspn(ip,"\n")]=0;
            snprintf(port,8,"2222");
            /* ensure sshd running */
            (void)!system("pgrep -x sshd >/dev/null||sudo service ssh start");
            /* check/setup port forward */
            char pf[B];pcmd("powershell.exe -c 'netsh interface portproxy show all' 2>/dev/null",pf,B);
            if(!strstr(pf,"2222")){
                char wip[128];pcmd("hostname -I 2>/dev/null|awk '{printf $1}'",wip,128);
                puts("Setting up Windows port forward (UAC)...");
                char c[B*2];snprintf(c,B*2,"powershell.exe -c \"Start-Process powershell -Verb RunAs -ArgumentList '-c',"
                    "'netsh interface portproxy delete v4tov4 listenport=2222 listenaddress=0.0.0.0 2>\\$null;"
                    "netsh interface portproxy add v4tov4 listenport=2222 listenaddress=0.0.0.0 connectport=22 connectaddress=%s;"
                    "netsh advfirewall firewall delete rule name=\\\"WSL SSH\\\" 2>\\$null;"
                    "netsh advfirewall firewall add rule name=\\\"WSL SSH\\\" dir=in action=allow protocol=tcp localport=2222'\"",wip);
                (void)!system(c);printf("Press Enter after admin window completes...");getchar();}
            printf("\xe2\x9c\x93 WSL port forward\n");
#ifdef __APPLE__
        }else if(1){
            pcmd("ipconfig getifaddr en0 2>/dev/null",ip,128);ip[strcspn(ip,"\n")]=0;
#else
        }else{
#endif
            pcmd("hostname -I 2>/dev/null|awk '{printf $1}'",ip,128);
            if(!ip[0])pcmd("ifconfig 2>/dev/null|awk '/inet /{if($2!~/^127/){printf $2;exit}}'",ip,128);
            if(!access("/data/data/com.termux",F_OK))snprintf(port,8,"8022");
            else{char pp[8];pcmd("awk '/^Port /{printf $2}' /etc/ssh/sshd_config 2>/dev/null",pp,8);if(pp[0])snprintf(port,8,"%s",pp);}}
        snprintf(h,256,!strcmp(port,"22")?"%s@%s":"%s@%s:%s",u?u:"",ip,port);
        char os[128];pcmd("uname -sr 2>/dev/null",os,128);os[strcspn(os,"\n")]=0;
        /* preserve existing password */
        const char*epw=NULL;for(int i=0;i<nh;i++)if(!strcmp(H[i].name,nm)){epw=H[i].pw;break;}
        ssh_savex(dir,nm,h,epw,"OS",os);printf("\xe2\x9c\x93 %s %s [%s]\n",nm,h,os);return 0;}
    /* rm */
    if(!strcmp(sub,"rm")&&argc>3){int x=-1;const char*a=argv[3];
        if(isdigit((unsigned char)*a))x=atoi(a);
        else{for(int i=0;i<nh;i++)if(!strcmp(H[i].name,a)){x=i;break;}}
        if(x>=0&&x<nh){char f[P];snprintf(f,P,"%s/%s.txt",dir,H[x].name);unlink(f);sync_repo();printf("\xe2\x9c\x93 rm %s\n",H[x].name);}return 0;}
    /* pw — change password */
    if(!strcmp(sub,"pw")&&argc>3){int x=-1;const char*a=argv[3];
        if(isdigit((unsigned char)*a))x=atoi(a);
        else{for(int i=0;i<nh;i++)if(!strcmp(H[i].name,a)){x=i;break;}}
        if(x>=0&&x<nh){char pw[256];printf("Password for %s: ",H[x].name);
            if(fgets(pw,256,stdin)){pw[strcspn(pw,"\n")]=0;ssh_save(dir,H[x].name,H[x].host,pw);printf("\xe2\x9c\x93 %s\n",H[x].name);}}return 0;}
    /* mv/rename */
    if((!strcmp(sub,"mv")||!strcmp(sub,"rename"))&&argc>4){
        const char*old=argv[3],*nn=argv[4];int x=-1;
        if(isdigit((unsigned char)*old))x=atoi(old);
        else{for(int i=0;i<nh;i++)if(!strcmp(H[i].name,old)){x=i;break;}}
        if(x>=0&&x<nh){char f[P];snprintf(f,P,"%s/%s.txt",dir,H[x].name);unlink(f);
            ssh_save(dir,nn,H[x].host,H[x].pw);printf("\xe2\x9c\x93 %s -> %s\n",H[x].name,nn);}return 0;}
    /* info */
    if(!strcmp(sub,"info")||!strcmp(sub,"i")){
        for(int i=0;i<nh;i++){char hp[256],port[8];ssh_parse(H[i].host,hp,port);
            printf("%s: ssh %s%s%s%s\n",H[i].name,strcmp(port,"22")?"-p ":"",strcmp(port,"22")?port:"",strcmp(port,"22")?" ":"",hp);}return 0;}
    /* os — detect remote OS on all hosts */
    if(!strcmp(sub,"os")){
        struct{int fd;pid_t pid;char nm[128],host[256],pw[256];}S[32];int ns=0;
        for(int i=0;i<nh&&ns<32;i++){int pfd[2];if(pipe(pfd))continue;
            pid_t p=fork();if(p==0){close(pfd[0]);char hp[256],port[8];ssh_parse(H[i].host,hp,port);
                char c[B*2];int l=ssh_pre(c,(int)sizeof c,H[i].pw,"-oConnectTimeout=5 -oStrictHostKeyChecking=no",port,hp);
                snprintf(c+l,(size_t)(sizeof(c)-(size_t)l)," 'uname -sr' 2>&1");
                char o[128];int r=pcmd(c,o,128);o[strcspn(o,"\n")]=0;
                if(!r&&o[0])(void)!write(pfd[1],o,strlen(o));close(pfd[1]);_exit(0);}
            close(pfd[1]);snprintf(S[ns].nm,128,"%s",H[i].name);snprintf(S[ns].host,256,"%s",H[i].host);
            snprintf(S[ns].pw,256,"%s",H[i].pw);S[ns].fd=pfd[0];S[ns].pid=p;ns++;}
        for(int i=0;i<ns;i++){char o[128];int l=(int)read(S[i].fd,o,127);o[l>0?l:0]=0;close(S[i].fd);waitpid(S[i].pid,NULL,0);
            if(o[0]){ssh_savex(dir,S[i].nm,S[i].host,S[i].pw,"OS",o);printf("\xe2\x9c\x93 %s: %s\n",S[i].nm,o);}
            else printf("x %s\n",S[i].nm);}
        return 0;}
    /* all/broadcast — parallel */
    if((!strcmp(sub,"all")||!strcmp(sub,"*"))&&argc>3){
        char cmd[B]="";for(int i=3,l=0;i<argc;i++)l+=snprintf(cmd+l,(size_t)(B-l),"%s%s",i>3?" ":"",argv[i]);
        char qc[B];snprintf(qc,B," 'bash -c '\"'\"'export PATH=$HOME/.local/bin:$PATH; %s'\"'\"'' 2>&1",cmd);
        struct{int fd;pid_t pid;char nm[128];}S[32];int ns=0;
        for(int i=0;i<nh&&ns<32;i++){int pfd[2];if(pipe(pfd))continue;
            pid_t p=fork();if(p==0){close(pfd[0]);char hp[256],port[8];ssh_parse(H[i].host,hp,port);
                char c[B*2];int l=ssh_pre(c,(int)sizeof c,H[i].pw,"-oConnectTimeout=5 -oStrictHostKeyChecking=no",port,hp);
                snprintf(c+l,(size_t)(sizeof(c)-(size_t)l),"%s",qc);char o[B];int r=pcmd(c,o,B);
                l=snprintf(c,B,"%c%s",r?'x':'+',o);(void)!write(pfd[1],c,(size_t)l);close(pfd[1]);_exit(0);}
            close(pfd[1]);snprintf(S[ns].nm,128,"%s",H[i].name);S[ns].fd=pfd[0];S[ns].pid=p;ns++;}
        for(int i=0;i<ns;i++){char o[B];int l=(int)read(S[i].fd,o,B-1);o[l>0?l:0]=0;close(S[i].fd);waitpid(S[i].pid,NULL,0);
            printf("\n%s %s\n",o[0]=='+'?"\xe2\x9c\x93":"x",S[i].nm);if(o[1])printf("%s",o+1);}
        return 0;}
    /* resolve host by # or name */
    int idx=-1;
    if(isdigit((unsigned char)*sub))idx=atoi(sub);
    else{for(int i=0;i<nh;i++)if(!strcmp(H[i].name,sub)){idx=i;break;}}
    if(idx<0||idx>=nh){printf("x No host %s\n",sub);return 1;}
    char hp[256],port[8];ssh_parse(H[idx].host,hp,port);
    if(!H[idx].pw[0]){char tc[B];int l=ssh_pre(tc,B,"","-oBatchMode=yes -oConnectTimeout=3",port,hp);
        snprintf(tc+l,(size_t)(B-l)," true 2>/dev/null");
        if(system(tc)){char pw[256];printf("Password for %s: ",H[idx].name);
            if(fgets(pw,256,stdin)){pw[strcspn(pw,"\n")]=0;if(pw[0]){snprintf(H[idx].pw,256,"%s",pw);ssh_save(dir,H[idx].name,H[idx].host,pw);}}}}
    {char cmd[B]="",c[B*2];for(int i=3;i<argc;i++){int l=(int)strlen(cmd);snprintf(cmd+l,(size_t)(B-l),"%s%s",l?" ":"",argv[i]);}
        int n=ssh_pre(c,(int)sizeof c,H[idx].pw,"-tt -oConnectTimeout=5 -oStrictHostKeyChecking=accept-new",port,hp);
        if(cmd[0])snprintf(c+n,(size_t)(sizeof(c)-(size_t)n)," 'bash -c '\"'\"'export PATH=$HOME/.local/bin:$PATH; %s'\"'\"''",cmd);
        else printf("Connecting to %s...\n",H[idx].name);
        execl("/bin/sh","sh","-c",c,(char*)NULL);_exit(127);}
}
